torch/csrc/jit/codegen/cuda/transform_replay.cpp

```diff
@@ -687,134 +687,447 @@ std::deque<TensorView*> producersFor(TensorView* tv) {
+size_t getReplayPos(const NextHopInfo& next_hop) {
+  auto& root_id_info = next_hop.root_id_info_from;
+  auto from_tv = next_hop.from;
+  // Flatten `root_id_info_from` to get the list of ids in the `from_tv`'s
+  // root/rfactor domain that contains information about the reference tensor.
+  std::unordered_set<IterDomain*> from_ids;
+  from_ids.reserve(root_id_info.size());
+  for (auto info : root_id_info) {
+    for (auto id : info.mapped_ids) {
+      from_ids.insert(id);
     }
   }
-
-  producer_tv->setDomain(replayed_producer.first);
-  replayed_pos[producer_tv] = replayed_producer.second;
-  n_replayed_root_dims[producer_tv] = n_transformed_root_dims;
-
-  return true;
+  // Get leaf IDs that contain information of `from_ids`
+  std::unordered_set<IterDomain*> relevant_leaves;
+  std::vector<IterDomain*> to_visit(from_ids.begin(), from_ids.end());
+  while (!to_visit.empty()) {
+    auto front = to_visit.back();
+    to_visit.pop_back();
+    if (front->uses().empty()) {
+      relevant_leaves.emplace(front);
+    } else {
+      for (auto def : front->uses()) {
+        auto outs = ir_utils::filterByType<IterDomain>(def->outputs());
+        to_visit.insert(to_visit.end(), outs.begin(), outs.end());
+      }
+    }
+  }
+  // Find the pos where all leaf IDs at <= pos contains
+  // information about the starting root domain
+  //
+  // TODO: should I change to the following behavior?
+  //
+  // Find the smallest pos where all leaf IDs containing
+  // information about the starting root domain are <= pos
+  //
+  // For example, if I have
+  //   preserved root domain: [I1, I2, I3, I4]
+  //   leaf domain: [I5, I6, I7, I8]
+  // where
+  //   I5 = merge(I1, I2)
+  //   I6 = something unrelated
+  //   I7 = merge(I3, I4)
+  //   I8 = something unrelated
+  // should I return 1, or 3 ?
+
+  // size_t i;
+  // for (i = from_tv->nDims() - 1; i >= 0; i--) {
+  //   if (relevant_leaves.count(from_tv->axis(i)) > 0) {
+  //     break;
+  //   }
+  // }
+  // return i + 1;
+
+  for (size_t i = 0; i < from_tv->nDims(); i++) {
+    if (relevant_leaves.count(from_tv->axis(i)) == 0) {
+      return i;
+    }
+  }
+  return from_tv->nDims();
 }
 
-bool TransformPropagator::replayCasP(
-    TensorView* consumer_tv,
-    TensorView* producer_tv) {
-  if (consumer_tv == starting_tv) {
-    return false;
+// Given `root_ids`, a list of IDs in the root domain of `tv`, find their
+// corresponding IDs in the rfactor domain of `tv`.
+std::unordered_set<IterDomain*> mapRootToRFactor(
+    TensorView* tv,
+    const std::unordered_set<IterDomain*>& root_ids) {
+  std::unordered_set<IterDomain*> mapped_rfactor_ids;
+  const auto& rfactor_dom = tv->getMaybeRFactorDomain();
+  for (auto id : rfactor_dom) {
+    if (root_ids.count(id) > 0) {
+      mapped_rfactor_ids.emplace(id);
+      continue;
+    }
+    for (auto root_id : root_ids) {
+      if (id == root_id || DependencyCheck::isDependencyOf(root_id, id)) {
```

Comment:
So this is giving full credit to a domain in rfactor that could potentially only have a partial domain from the root. I wonder how safe this is through complex uses of view. Are there instances where we would have to accurately track "partial" ownership of an rfactor domain with view?

Reply:
That's an interesting question. I don't know if there's an actually adverse case. It seems to me also about how propagation should be done, whether transformation should be propagated through partial ownership? Not sure which should be preferred.